### PR TITLE
fix producing multiple identical opaque pointer types

### LINKED ISSUE #79057: [SPIRV] SPIR-V Backend crashes on creation of a struct with opaque pointer fields
An attempt to use opaque pointer fields in a struct makes SPIR-V Backend crash on the stage of module analysis if the struct has more than one opaque pointer field. For example:

```
@a = addrspace(1) constant i32 123
@struct = addrspace(1) global {ptr addrspace(1), ptr addrspace(1)} { ptr addrspace(1) @a, ptr addrspace(1) @a }

define spir_kernel void @foo() {
  ret void
}
```

demonstrates a crash in `SPIRVGeneralDuplicatesTracker::buildDepsGraph()` due to the fact that there is no mandatory records defined in `SPIRVReg2EntryTy Reg2Entry;` for some type references of `OpTypeStruct`: namely, for all opaque pointer type arguments of `OpTypeStruct` after the first one.

The root cause is producing of multiple identical opaque pointer types, i.e., creation of new `OpTypePointer` without checking their existence. Along with inserting just a single record per type into a data structure used for dependencies analysis in `SPIRVGeneralDuplicatesTracker::buildDepsGraph()` this causes missing records in the same data structure for all opaque pointer fields except for the first one.

PR: https://github.com/llvm/llvm-project/pull/79060

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR fixes https://github.com/llvm/llvm-project/issues/79057 and improves code generation for opaque pointers by replacing the culprit SPIRVGlobalRegistry::getOpTypePointer() call with a more appropriate SPIRVGlobalRegistry::getOrCreateSPIRVPointerType() call. The latter function works together with the `DuplicatesTracker` (`SPIRVGeneralDuplicatesTracker DT;` from `class SPIRVGlobalRegistry`) to trace existence of previous definitions of opaque pointers. This allows to produce just one `OpTypePointer` command for all identical opaque pointers definitions and to return the very same type record for subsequent `SPIRVGlobalRegistry::createSPIRVType()` invocations.

This PR alone improves code generation by producing a single needed definition per all opaque pointers to i8 of the same address space instead of multiple identical definitions produced before the patch. From the root cause analysis of https://github.com/llvm/llvm-project/issues/79057 we see also that this PR resolves the problem of inconsistency between keeping multiple instruction for identical opaque pointer types and just a single record for all such instructions in the `DuplicatesTracker`, and so it also resolves the issue with crashes on creation of a struct with opaque pointer fields due to the fact that now such struct fields refer to the same operand `<id>` having a required record in the data structure used for dependencies analysis (see https://github.com/llvm/llvm-project/issues/79057).

### CODE REVIEW

**Comment by VyacheslavLevytskyy** [comment]
@michalpaszkowski a simple fix for a nasty problem

File: llvm/test/CodeGen/SPIRV/struct_opaque_pointers.ll [hunk 1]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc -O0 -mtriple=spirv32-unknown-unknown %s -o - | FileCheck %s
+
```
Comment (michalpaszkowski) [inline llvm/test/CodeGen/SPIRV/struct_opaque_pointers.ll:2]: Could spirv-val run line be added here as well?

**Review by michalpaszkowski — APPROVED** [review]
Thanks for the pull request! LGTM!

File: llvm/test/CodeGen/SPIRV/struct_opaque_pointers.ll [hunk 2]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc -O0 -mtriple=spirv32-unknown-unknown %s -o - | FileCheck %s
```
Comment (michalpaszkowski) [inline llvm/test/CodeGen/SPIRV/struct_opaque_pointers.ll:1]: Please also move the test to the `/pointers` directory and change underscores to dashes in the name.

File: llvm/test/CodeGen/SPIRV/struct_opaque_pointers.ll (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply llvm/test/CodeGen/SPIRV/struct_opaque_pointers.ll:2]: As in the case of https://github.com/llvm/llvm-project/pull/78293, adding spirv-val run line would make the test fail. The context is that both SPIR-V Backend and SPIR-V Translator do not comply with specification in the part that the constituents of OpConstantComposite must all be non-specialization constant-instruction declarations or OpUndef, with the difference that spirv-val tolerates SPIR-V Translator's wrapping of operands in OpSpecConstantComposite.
I'd suggest to postpone this to be a separate issue if it's acceptable, as in the case of #78293.

**Pull request merged**